source/D2Game/src/AI/AiBaal.cpp

```diff
@@ -96,6 +96,320 @@ void __fastcall AITHINK_Fn135_BaalCrab(D2GameStrc* pGame, D2UnitStrc* pUnit, D2A
+	int32_t bInFarRange = 0;
+	if (pAiCmd)
+	{
+		// UBER TWEAK START
+		const int32_t nDistance = AIUTIL_GetDistanceToCoordinates_FullUnitSize(pTarget, pUnit);
+		// UBER TWEAK END
+		if (nDistance > 75)
+		{
+			bInMediumRange = 1;
+		}
+
+		if (nDistance > 100)
+		{
+			bInFarRange = 1;
+		}
+	}
+
+	const int32_t bInMeleeRange = UNITS_IsInMeleeRange(pUnit, pTarget, 0);
+	const int32_t bNotCollidingWithWall = UNITS_TestCollisionWithUnit(pUnit, pTarget, COLLIDE_MISSILE_BARRIER) == 0;
+	if (bInMeleeRange)
+	{
+		// UBER TWEAK START
+		int32_t aiParams[16] = { 0, 0, 0, 0, 0, 0, 0, 0, 20, 10, 150, 10, 10, 70, 0, 0 };
+		// UBER TWEAK END
+
+		if (UNITS_GetCurrentLifePercentage(pTarget) < 33)
+		{
+			aiParams[10] += 50;
+		}
+
+		aiParams[1] += 100 - UNITS_GetCurrentLifePercentage(pUnit);
+
+		if (STATES_CheckState(pTarget, STATE_COLD))
+		{
+			aiParams[8] = 0;
+			aiParams[13] = 0;
+			aiParams[11] = 0;
+		}
+
+		if (!bNotCollidingWithWall)
+		{
+			aiParams[11] = 0;
+			aiParams[13] = 0;
+			aiParams[12] = 0;
+		}
+
+		return AI_GetRandomArrayIndex(aiParams, std::size(aiParams), pUnit, 1);
+	}
+
+	if (bNotCollidingWithWall)
+	{
+		return AIBAAL_RollRandomUberAiParamForNonCollidingUnit(pGame, pAiControl, pUnit, pTarget, nCount, bInMediumRange, bInFarRange, bInCloseRange, nMax);
+	}
+
+	// UBER TWEAK START
+	int32_t aiParams[16] = { 0, 0, 50, 0, 0, 0, 20, 0, 10, 10, 0, 0, 0, 0, 100, 0 };
+	// UBER TWEAK END
+
+	if (nCount < 2)
+	{
+		// UBER TWEAK START
+		aiParams[2] = 75;
+		// UBER TWEAK END
+		aiParams[10] = 25;
+	}
+
+	// UBER TWEAK START
+	if (AI_CheckSpecialSkillsOnPrimeEvil(pTarget))
+	{
+		if (!bInMediumRange)
+		{
+			aiParams[14] += 50;
+			aiParams[13] += 50;
+		}
+	}
+
+	if (bInMediumRange)
+	{
+		aiParams[2] = 0;
+		aiParams[6] = 0;
+		aiParams[14] += 100;
+		aiParams[11] += 25;
+		aiParams[13] += 25;
+	}
+
+	if (bInFarRange)
+	{
+		aiParams[14] += 100;
+	}
+	// UBER TWEAK END
+
+	return AI_GetRandomArrayIndex(aiParams, std::size(aiParams), pUnit, 1);
+}
+
+//1.14d: 0x005FD0F0
+void __fastcall AIBAAL_SpawnUberBaalMinion(D2GameStrc* pGame, D2UnitStrc* pUnit)
+{
+	D2CoordStrc coord;
+	coord.nX = CLIENTS_GetUnitX(pUnit);
+	coord.nY = CLIENTS_GetUnitY(pUnit);
+
+	const int32_t nMinionTypes[3] = { MONSTER_WRAITH9, MONSTER_VAMPIRE9, MONSTER_WILLOWISP8 };
+	const int32_t nMinionModes[3] = { MONMODE_NEUTRAL, MONMODE_CAST, MONMODE_SKILL1 };
+	// Note: doesn't pick willow wisps (they thought it was too hard?)
+	// change the 2 to 3 to enable them
+	int32_t nChoice = ITEMS_RollLimitedRandomNumber(&pUnit->pSeed, 2);
+
+	D2ActiveRoomStrc* pRoom = COLLISION_GetFreeCoordinates(UNITS_GetRoom(pUnit), &coord, 2, 0x100, 0);
+	if (pRoom)
+	{
+		D2UnitStrc* pMinion = D2GAME_SpawnMonster_6FC69F10(pGame, pRoom, coord.nX, coord.nY, nMinionTypes[nChoice], nMinionModes[nChoice], -1, 0);
+		if (pMinion)
+		{
+			pMinion->dwFlags |= UNITFLAG_NOXP; // Added in 1.13
+			STATES_ToggleState(pMinion, 184, 1);
+		}
+	}
+}
+
+//1.14d: 0x005FD200
```

Comment:
This function is a noop in my version, what is the source of this function? What does "UBER TWEAK START" mean?